libs/core/async_cuda/include/hpx/async_cuda/cuda_event.hpp

```diff
@@ -29,56 +31,77 @@ namespace hpx { namespace cuda { namespace experimental {
 
         // create a bunch of events on initialization
         cuda_event_pool()
-          : free_list_(initial_events_in_pool)
+          : max_number_devices_(0)
         {
-            for (int i = 0; i < initial_events_in_pool; ++i)
+            check_cuda_error(cudaGetDeviceCount(&max_number_devices_));
+            HPX_ASSERT_MSG(max_number_devices_ > 0,
+                "CUDA polling enabled and called, yet no CUDA device found!");
+            for (int device = 0; device < max_number_devices_; device++)
             {
-                add_event_to_pool();
+                check_cuda_error(cudaSetDevice(device));
+                free_lists_.emplace_back(initial_events_in_pool);
+                for (int i = 0; i < initial_events_in_pool; ++i)
+                {
+                    add_event_to_pool(device);
+                }
             }
         }
 
         // on destruction, all objects in stack will be freed
         ~cuda_event_pool()
         {
-            cudaEvent_t event;
-            bool ok = true;
-            while (ok)
+            HPX_ASSERT_MSG(free_lists_.size != max_number_devices_,
+                "Number of CUDA event pools does not match the number of "
+                "devices!");
+            for (int device = 0; device < max_number_devices_; device++)
             {
-                ok = free_list_.pop(event);
-                if (ok)
-                    check_cuda_error(cudaEventDestroy(event));
+                check_cuda_error(cudaSetDevice(device));
+                cudaEvent_t event;
+                bool ok = true;
+                while (ok)
+                {
+                    ok = free_lists_[device].pop(event);
+                    if (ok)
+                        check_cuda_error(cudaEventDestroy(event));
+                }
             }
         }
 
-        inline bool pop(cudaEvent_t& event)
+        inline bool pop(cudaEvent_t& event, int device)
```

Comment:
```suggestion
        inline bool pop(cudaEvent_t& event, int device = 1)
```
?

I'm not sure about the default value, though - does the CUDA device indexing start with `0` or `1`?